### PR TITLE
Multi-Core XDP Polling Support

### PR DESCRIPTION
**Pull request by nibanks**
## Description

Updates the XDP datapath to support creating multiple workers to poll on all configured CPUs.

## Testing

Existing tests will verify the default, single-core polling. New tests need to be added (once full test support is onboarded) for multi-core polling. Validated with manual testing on Azure VMs.

## Documentation

N/A

### CODE REVIEW

File: src/platform/datapath_raw_xdp.c [hunk 1]
```diff
@@ -66,26 +67,41 @@ typedef struct XDP_INTERFACE {
...
+} XDP_WORKER;
+
+void XdpWorkerAddQueue(_In_ XDP_WORKER* Worker, _In_ XDP_QUEUE* Queue) {
+    XDP_QUEUE** Tail = &Worker->Queues;
+    while (*Tail != NULL) {
+        Tail = &(*Tail)->Next;
+    }
+    *Tail = Queue;
+    Queue->Next = NULL;
+}
```
Comment (csujedihy) [inline src/platform/datapath_raw_xdp.c:87]: Is there a specific reason for not using `CXPLAT_LIST_ENTRY`?

File: src/platform/datapath_raw_xdp.c [hunk 2]
```diff
@@ -66,26 +67,41 @@ typedef struct XDP_INTERFACE {
...
 
-typedef struct XDP_DATAPATH {
-    CXPLAT_DATAPATH;
-
-    BOOLEAN Running;
+typedef struct QUIC_CACHEALIGN XDP_WORKER {
+    const struct XDP_DATAPATH* Xdp;
     HANDLE CompletionEvent;
+    XDP_QUEUE* Queues; // A linked list of queues, accessed by Next.
+    uint16_t Index;
```
Comment (csujedihy) [inline src/platform/datapath_raw_xdp.c:77]: >Index [](http://example.com/codeflow?start=13&length=5)

Can we name it ProcIndex?

File: src/platform/datapath_raw_xdp.c [hunk 3]
```diff
@@ -923,20 +930,37 @@ CxPlatDpRawInterfaceRemoveRules(
     CxPlatLockRelease(&Interface->RuleLock);
 }
 
+
```
Comment (csujedihy) [inline src/platform/datapath_raw_xdp.c:933]: delete ws

File: src/platform/datapath_raw_xdp.c [hunk 4]
```diff
@@ -923,20 +930,37 @@ CxPlatDpRawInterfaceRemoveRules(
...
+    _In_opt_ const CXPLAT_DATAPATH_CONFIG* Config
     )
 {
     XDP_DATAPATH* Xdp = (XDP_DATAPATH*)Datapath;
     QUIC_STATUS Status;
 
-    CxPlatXdpReadConfig(Xdp, Config);
-    CxPlatDpRawGenerateCpuTable(Datapath);
+    uint16_t DefaultCore = (uint16_t)(CxPlatProcMaxCount() - 1);
+    const uint16_t* PollingCores =
```
Comment (csujedihy) [inline src/platform/datapath_raw_xdp.c:957]: >PollingCores [](http://example.com/codeflow?start=20&length=12)

It seems that the term processor/proc is more widely used in QUIC and XDP than cores.

File: src/platform/datapath_raw_xdp.c (hunk 1 above)
Comment (nibanks) [reply src/platform/datapath_raw_xdp.c:87]: I don't need a doubly-linked list.

File: src/platform/datapath_raw_xdp.c (hunk 2 above)
Comment (nibanks) [reply src/platform/datapath_raw_xdp.c:77]: Sure

File: src/perf/lib/SecNetPerfMain.cpp [hunk 5]
```diff
@@ -87,8 +87,10 @@ PrintHelp(
         "  -cibir:<hex_bytes>          A CIBIR well-known idenfitier.\n"
         "\n"
         "Client: secnetperf -TestName:<Throughput|RPS|HPS> [options]\n"
+#ifndef _KERNEL_MODE
         "Both:\n"
         "  -cpu:<cpu_index>            Specify a processor for raw datapath thread(s) to run on.\n"
```
Comment (maolson-msft) [inline src/perf/lib/SecNetPerfMain.cpp:92]: Elsewhere you removed "raw" from the terminology; maybe you want to remove it here too.

Comment (nibanks) [reply src/perf/lib/SecNetPerfMain.cpp:92]: Sure. Let me know if you see any other places I missed.

File: src/platform/datapath_raw_xdp.c [hunk 6]
```diff
@@ -66,26 +67,41 @@ typedef struct XDP_INTERFACE {
...
+    XDP_QUEUE* Queues; // A linked list of queues, accessed by Next.
+    uint16_t ProcIndex;
+} XDP_WORKER;
+
+void XdpWorkerAddQueue(_In_ XDP_WORKER* Worker, _In_ XDP_QUEUE* Queue) {
+    XDP_QUEUE** Tail = &Worker->Queues;
+    while (*Tail != NULL) {
+        Tail = &(*Tail)->Next;
+    }
+    *Tail = Queue;
```
Comment (maolson-msft) [inline src/platform/datapath_raw_xdp.c:85]: Is there some particular reason the new queue is added to the tail instead of the head?

Comment (nibanks) [reply src/platform/datapath_raw_xdp.c:85]: Not really, I guess. I just tried to keep them in order that I created them, but I don't have any good justification requiring that.

File: src/platform/datapath_raw_dpdk.c [hunk 7]
```diff
@@ -144,9 +144,10 @@ CxPlatDpdkReadConfig(
 _IRQL_requires_max_(PASSIVE_LEVEL)
 size_t
 CxPlatDpRawGetDapathSize(
```
Comment (maolson-msft) [inline src/platform/datapath_raw_dpdk.c:146]: Maybe go ahead and fix the spelling of "Dapath" here

File: src/platform/datapath_raw_xdp.c [hunk 8]
```diff
@@ -742,6 +740,15 @@ CxPlatDpRawInterfaceInitialize(
         }
     }
 
+    //
+    // Add the queues to a worker (round robin).
```
Comment (maolson-msft) [inline src/platform/datapath_raw_xdp.c:744]:
```suggestion
    // Add each queue to a worker (round robin).
```

**Review by thhous-msft — APPROVED** [review]

**Pull request merged**